File: 2019-2020_programmers/_06_/src/kotlin/Camouflage.kt
val cArr: List<List<List<String>>> 
    = listOf(
        listOf(listOf("yellow_hat", "headgear"), listOf("blue_sunglasses", "eyewear"), listOf("green_turban", "headgear")),
        listOf(listOf("crow_mask", "face"), listOf("blue_sunglasses", "face"), listOf("smoky_makeup", "face"))
    )

fun solution(clothes: List<List<String>>):Int{
    val category = mutableMapOf<String, MutableList<String>>()
    val nums = mutableListOf<Int>()

    for (c in clothes) {
        val key = c[1]
        if (category.get(key) == null) {
            val l = mutableListOf<String>()
            l.add(c[0])
            category[key] = l 
        } else {
            val l: MutableList<String> = category.get(key)!!
            l.add(c[0])
        }
    }

    for (k in category.keys) {
        nums.add(category[k]!!.size)
    }

    if (nums.size == 1) {
        return nums[0]
    } else {
        return nums.fold(1) {x:Int, y:Int -> x * (y + 1)} - 1
    }
}

fun main(args: Array<String>) {
    println(solution(cArr[0]))
}

File: 2019-2020_programmers/_06_/src/python/camouflage.py
getCArr = [
    [["yellow_hat", "headgear"], ["blue_sunglasses", "eyewear"], ["green_turban", "headgear"]],
    [["crow_mask", "face"], ["blue_sunglasses", "face"], ["smoky_makeup", "face"]]
]

from functools import reduce

def solution(clothes):
    category = {}
    nums = []

    for c in clothes:
        key = c[1]
        if category.get(key, None) is None:
            l = []; l.append(c[0])
            category[key] = l
        else:
            l = category.get(key)
            l.append(c[0])

    for k in category.keys():
        nums.append(len(category[k]))
    # sum = 1
    if len(nums) == 1:
        return len(nums)
    else:
        # for n in nums:
        #     sum = sum * (n + 1)
        # return sum - 1
        return reduce(lambda x,y: x * (y + 1), nums, 1) - 1

print(solution(getCArr[0]))
